src/libraries/Hooks.sol

```diff
@@ -148,24 +191,50 @@ library Hooks {
         IPoolManager.ModifyLiquidityParams memory params,
         BalanceDelta delta,
         bytes calldata hookData
-    ) internal {
-        if (params.liquidityDelta > 0 && key.hooks.hasPermission(AFTER_ADD_LIQUIDITY_FLAG)) {
-            self.callHook(
-                abi.encodeWithSelector(IHooks.afterAddLiquidity.selector, msg.sender, key, params, delta, hookData)
-            );
-        } else if (params.liquidityDelta <= 0 && key.hooks.hasPermission(AFTER_REMOVE_LIQUIDITY_FLAG)) {
-            self.callHook(
-                abi.encodeWithSelector(IHooks.afterRemoveLiquidity.selector, msg.sender, key, params, delta, hookData)
-            );
+    ) internal returns (BalanceDelta hookDelta) {
+        if (params.liquidityDelta > 0) {
+            if (key.hooks.hasPermission(AFTER_ADD_LIQUIDITY_FLAG)) {
+                hookDelta = BalanceDelta.wrap(
+                    self.callHookWithReturnDelta(
+                        abi.encodeWithSelector(
+                            IHooks.afterAddLiquidity.selector, msg.sender, key, params, delta, hookData
+                        ),
+                        key.hooks.hasPermission(AFTER_ADD_LIQUIDITY_RETURNS_DELTA_FLAG)
+                    )
+                );
+            }
+        } else {
+            if (key.hooks.hasPermission(AFTER_REMOVE_LIQUIDITY_FLAG)) {
+                hookDelta = BalanceDelta.wrap(
+                    self.callHookWithReturnDelta(
+                        abi.encodeWithSelector(
+                            IHooks.afterRemoveLiquidity.selector, msg.sender, key, params, delta, hookData
+                        ),
+                        key.hooks.hasPermission(AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA_FLAG)
+                    )
+                );
+            }
         }
     }
 
     /// @notice calls beforeSwap hook if permissioned and validates return value
     function beforeSwap(IHooks self, PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
         internal
+        returns (int256 amountToSwap, int128 hookDeltaInSpecified)
     {
+        amountToSwap = params.amountSpecified;
         if (key.hooks.hasPermission(BEFORE_SWAP_FLAG)) {
-            self.callHook(abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData));
+            hookDeltaInSpecified = self.callHookWithReturnDelta(
+                abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData),
+                key.hooks.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG)
+            ).toInt128();
+
+            // Update the swap amount according to the hook's return, and check that the swap type doesnt change (exact input/output)
+            if (hookDeltaInSpecified != 0) {
+                bool exactInput = amountToSwap < 0;
+                amountToSwap += hookDeltaInSpecified;
```

Comment:
Reading this line of code I feel like I need to run through every single case of swap type, hook delta return type, and make sure that the amountToSwap calculation lines up with what the user pays or takes, and check that that also lines up with the respective deltas applied, and that it also aligns with hook pays or takes, and that that then also aligns with the hookDeltas. Is there a simpler way to reason about this and check this? I think in general, maybe each of these cases just needs to have a respective test, and documentation could live there?

Here are all the cases I'm thinking about:

- exactOutput swap, hook applies a negative delta on specified
- exactOutput swap, hook applies a positive delta on specified
- exactInput swap, hook applies a negative delta on specified
- exactInput swap, hook applies a positive delta on specified

And then each case applies for the unspecified one as well.

I think we should just make sure we have test case for each one and proper documentation for what we expect to happen for each one, could live in README or test cases.

Reply:
Running through a couple of these, it seems to line up with my mental model, but understanding each of these is definitely not easy to reason through

Reply:
have now added a fuzz test that hopefully makes it all feel better, but yes we should write good docs on it all